fix: sqlsrv default port cause connection refuse

puschie286 opened the pull request:
**Description**
database connection is refused when using sqlsrv driver with default port. 
fixed by ignoring the default port ( 1433 )

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

Fixes: https://github.com/codeigniter4/CodeIgniter4/issues/6165

datamweb commented:
All code must conform to our [Style Guide](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/styleguide.md), which is based on PSR-12.
You can fix most of the coding style violations by running this command in your terminal:
```console
composer cs-fix
```
more info [here](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#php-style).

kenjis commented:
What is the issue?
#6165 is not a bug.

We have tests with SQLSRV port 1433, and all tests pass.
https://github.com/codeigniter4/CodeIgniter4/blob/1e7204bbbacc2731327d79295bc32735f28b548f/tests/_support/Config/Registrar.php#L85-L102

puschie286 commented:
tested it on a windows server 2019 with default (ms sql server) installation - all database test fail with the following exception:

`CodeIgniter\Database\Exceptions\DatabaseException : Unable to connect to the database.
Main connection [SQLSRV]: [Microsoft][ODBC Driver 17 for SQL Server]TCP Provider: No connection could be made because the target machine actively refused it.
 SQLSTATE: 08001, code: 10061
[Microsoft][ODBC Driver 17 for SQL Server]Login timeout expired SQLSTATE: HYT00, code: 0
[Microsoft][ODBC Driver 17 for SQL Server]A network-related or instance-specific error has occurred while establishing a connection to SQL Server. Server is not found or not accessible. Check if instance name is correct and if SQL Server is configured to allow remote connections. For more information see SQL Server Books Online. SQLSTATE: 08001, code: 10061`

So its not valid to add the default port when executing on a windows platform. 
@kenjis 
To find problems like this, you need to add a test runner for windows platforms
( Beside there are many test that fail on windows server because of the line ending )

kenjis commented:
@puschie286 I got the error occurs only on Windows server.

I don't use Windows.
Contributions are welcome. See #7491

michalsn commented:
@puschie286 Stupid question... have you checked that the port number you are trying to use is correct? https://www.mssqltips.com/sqlservertip/2495/identify-sql-server-tcp-ip-port-being-used/

puschie286 commented:
@michalsn we use the **default** configuration, that is dynamic port ( [Ref](https://learn.microsoft.com/en-us/troubleshoot/sql/database-engine/connect/static-or-dynamic-port-config) "By default, a SQL Server named instance is configured to listen on dynamic ports." ). this is probably the reason why the fixed port in the database config doesnt work with the default installation.

michalsn commented:
@puschie286 Ok, now I'm lost. If you're using dynamic ports, why are you expecting `1433` to work?

puschie286 commented:
@michalsn you're right, for dynamic port we should remove the port config entry. guess thats the easiest solution

my concern is backward compatibility, some older installation use static 1433 and new installation dynamic port - and both will work without a port definition. because the default config has the 'port' entry, you might think you need to define it and a quick search will give you the anwser that 1433 is the default port for ms sql server.

maybe a notice in the database configuration documentation would be enough to help users to configure sqlsrv driver for new installations.

for now we will apply a migration ( on our end ) that remove all sqlsrv port entries with the value of 1433 to have the backward compatibility with old installations

michalsn commented:
@puschie286 Ok, now I understand what you mean. Thanks.

I'm not a fan of the proposed change. I don't want us to end up with a situation where someone defines a port as `1433` and their base is actually running on a different port.

A much more appropriate approach would be to add a note in the user guide - just like you mentioned.

puschie286 commented:
@kenjis yes, dynamic port is the default setting nowadays.

A named instance is an installation with an different instance name than the default instance name - overall its just an alias to identify the sql server instance ( if you have multiple on the same host ). if you dont provide the instance name, the driver use the default instance name - so its not possible to have an "unnamed instance" ( this term is sometimes used but it just refer to default instance name ). You could say that the instance name is microsoft way of omit the port

i removed my change and added these information in the database configuration documentation

kenjis commented on user_guide_src/source/database/configuration.rst:
```diff
@@ -174,7 +174,7 @@ Explanation of Values:
 **strictOn**    true/false (boolean) - Whether to force "Strict Mode" connections, good for ensuring strict SQL
                 while developing an application (``MySQLi`` only).
-**port**        The database port number.
+**port**        The database port number - Empty for default port (or dynamic port with ``SQLSRV``)
```
Do you mean an empty string?

puschie286 replied:
yes - empty string is the only value that works for all driver 
( sqlsrv and postgre doesnt use empty function for validation )

kenjis commented on user_guide_src/source/database/configuration.rst:
```diff
@@ -174,7 +174,7 @@ Explanation of Values:
 **strictOn**    true/false (boolean) - Whether to force "Strict Mode" connections, good for ensuring strict SQL
                 while developing an application (``MySQLi`` only).
-**port**        The database port number.
+**port**        The database port number - Empty for default port (or dynamic port with ``SQLSRV``)
```
```suggestion
**port**        The database port number - Empty string ``''`` for default port (or dynamic port with ``SQLSRV``).
```

kenjis reviewed (approved):
Thank you!
